pkgs/top-level/unit-files.nix

```diff
@@ -0,0 +1,105 @@
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
+let
+
+  # We don't _need_ `lib` here, but importing it wouldn't be detrimental to performance either because of Nix's `import` cache
+  inherit (builtins) attrNames elemAt length mapAttrs readDir;
+
+  # Warning: Do not make this configurable, otherwise the `import` evaluation cache won't work!
+  baseDirectory = ../unit;
+
+  entries = readDir baseDirectory;
+  shards = attrNames entries;
+
+  # Return the name mappings in a single shard
+  # Type: String -> AttrsOf Path
+  namesForShard = shard:
+    if entries.${shard} != "directory" then
+      # Ignore non-directories. Technically only Readme.md is allowed as a file in the base directory, so we could alternatively:
+      # - Assume that Readme.md is the only file and change the condition to `shard == "Readme.md"` for a minor performance improvement.
+      #   This would however cause very poor error messages if there's other files.
+      # - Ensure that Readme.md is the only file, throwing a better error message if that's not the case.
+      #   However for good code architecture, all error messages relating to `pkgs/unit` should be centralized in (TODO: Link to validity checks)
+      # Additionally in either of those alternatives, we would have to duplicate the hardcoding of "Readme.md"
+      { }
+    else
+      mapAttrs
+        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
+        (readDir (baseDirectory + "/${shard}"));
+
+  # Merge the name mappings of all shards between the indices start (inclusive) and end (exclusive)
+  # This ignores any potential duplicates for efficiency
+  # Invariant: start < end
+  # Type: Int -> Int -> AttrsOf Path
+  binaryMerge = start: end:
+    if end - start >= 2 then
+      # If there's more than 2 elements, split the range in two, recurse on each part and merge the result
+      # The binaryMerge invariant is satisfied because each half will have at least 1 element
+      binaryMerge start (start + (end - start) / 2)
+      // binaryMerge (start + (end - start) / 2) end
+    else
+      # Otherwise there will be exactly 1 element due to the invariant, in which case we just return it directly
+      namesForShard (elemAt shards start);
+
+  # We could get the same result with a simple
+  #
+  #   builtins.foldl' (acc: entry: acc // namesForShard entry) {} shards;
+  #
+  # but this has rather poor evaluation performance for a large number of shards.
+  # This is because `l // r` has complexity O(l + r) with l and r being the number attributes on the left and right side respectively.
+  # So e.g. if we have n shards each with m entries, the above fold would be evaluated like this:
+  #     /\     O((n-1)m + m) = O(nm)
+  #    ⋰  m    ... (n times)
+  #   /\       O(2m + m)     = O(3m)
+  #  /\ m      O(m + m)      = O(2m)
```

Comment:
nit(mathematician): `O(m)` :p

Reply:
It's pretty clear that the point here is that the constants in O are «roughly» the same, and then we want the bottom to say 2m

Reply:
nit²: `O(m) = O(m + m)`, so so both are correct

Reply:
Only as long as your O(2m), O(3m)… O(nm) do not want to have the same constant.

Ideally it would even be `Θ(·)`, not `O(·)`…

Reply:
I mean, more precisely, summing $O(Km)$ with K constant is a bit clunky, this is a nit, but to the non-expert, it can give the impression that $\sum_k O(km) = O(m \sum_k k) = O(m(m + 1)/2) = O(m^2)$ I imagine.

I don't really care, just wanted to point out that replacing $O$ by a proper sequence $T(n, m)$ counting the number of elementary operations wrt to `//` might be enough, and you can conclude with any Landau notation that you prefer.

Reply:
However, I believe it is the case that $\sum_{k=1}^{n} \mathcal{O}(km) \subseteq \mathcal{O(n^2m)}$, which is used here. Though I admit that originally I did it with the triangle formula before simplifying :)

Reply:
Although, since the argument here is that the linear algorithm is slower, I'd have to prove that it's worse than the binary merge algorithm, so I believe to be fully correct I should calculate the $\Omega$ (a lower bound) for the linear algorithm and $\mathcal{O}$ (an upper bound) for the binary merge algorithm.

At which point just counting the operations like @RaitoBezarius suggested seems much easier!

Reply:
No, consider the almost finite sequence u_n^k = kn if n <= k else 0.
All u_n^k are O(kn) where k is a fixed constant.
But sum of u_n^k is not an O(n^2) because you cannot bound uniformly the constant for all k

(Example written quickly, I am not totally sure it works but it sure feels a counterexample)

Reply:
I see, thanks! I now changed it to count the number of copies directly, and only using $\mathcal{O}$ for the total